File: examples/easy_embed_demo.cpp
// ─── Easy Embed Demo (C++) ─────────────────────────────────────────────────────
//
// Demonstrates the spectra::easy_embed API for one-liner offscreen rendering.
// No windows, no event loop, no daemon — just data in → pixels out.
//
// Build:
//   cmake -DSPECTRA_BUILD_EMBED_SHARED=ON ..
//   g++ -std=c++20 -I include examples/easy_embed_demo.cpp -L build -lspectra -o easy_embed_demo
//
// Usage:
//   ./easy_embed_demo
//

#include <spectra/spectra.hpp>
#include <iostream>
#include <vector>
#include <cmath>

void demo_basic_line()
{
    std::cout << "📈 Basic line plot\n";
    std::vector<float> x = {0, 1, 2, 3, 4, 5};
    std::vector<float> y = {0, 1, 4, 9, 16, 25};  // y = x^2
    
    auto img = spectra::render(x, y);
    std::cout << "   Rendered " << img.width << "x" << img.height 
              << " image (" << img.size_bytes() << " bytes)\n";
    
    // Count non-zero pixels to verify something was rendered
    size_t nonzero = 0;
    for (uint8_t b : img.data)
        if (b != 0) ++nonzero;
    std::cout << "   Non-zero pixels: " << nonzero << "\n\n";
}

void demo_save_to_file()
{
    std::cout << "💾 Save to PNG file\n";
    std::vector<float> x = {0, 1, 2, 3, 4, 5};
    std::vector<float> y = {0, 1, 4, 9, 16, 25};
    
    spectra::RenderOptions opts;
    opts.save_path = "easy_embed_demo.png";
    auto img = spectra::render(x, y, opts);
    
    std::cout << "   Saved to " << opts.save_path << "\n\n";
}

void demo_custom_size()
{
    std::cout << "🖼️  Custom size and styling\n";
    std::vector<float> x = {0, 1, 2, 3, 4, 5};
    std::vector<float> y = {0, 1, 4, 9, 16, 25};
    
    spectra::RenderOptions opts;
    opts.width = 1920;
    opts.height = 1080;
    opts.fmt = "r--o";  // Red dashed line with circles
    
    auto img = spectra::render(x, y, opts);
    std::cout << "   HD render: " << img.width << "x" << img.height 
              << " with red dashed line and circles\n\n";
}

void demo_scatter()
{
    std::cout << "🔵 Scatter plot\n";
    std::vector<float> x, y;
    
    // Generate random scatter data
    std::srand(42);
    for (int i = 0; i < 100; ++i)
    {
        x.push_back(std::rand() / float(RAND_MAX) * 10.0f);
        y.push_back(std::rand() / float(RAND_MAX) * 10.0f);
    }
    
    spectra::RenderOptions opts;
    opts.save_path = "scatter_demo.png";
    auto img = spectra::render_scatter(x, y, opts);
    std::cout << "   Scatter plot with 100 points saved to scatter_demo.png\n\n";
}

void demo_multi_series()
{
    std::cout << "📊 Multi-series plot\n";
    std::vector<float> x;
    for (int i = 0; i < 100; ++i)
        x.push_back(i * 0.1f);
    
    std::vector<float> y1, y2, y3;
    for (float v : x)
    {
        y1.push_back(std::sin(v));
        y2.push_back(std::cos(v));
        y3.push_back(v * 0.5f);  // Linear
    }
    
    spectra::RenderOptions opts;
    opts.title = "Trigonometric Functions";
    opts.save_path = "multi_demo.png";
    
    auto img = spectra::render_multi({
        {x, y1, "b-", "sin(x)"},
        {x, y2, "r--", "cos(x)"},
        {x, y3, "g:", "0.5x"},
    }, opts);
    
    std::cout << "   Multi-series plot with sin, cos, and linear functions\n";
    std::cout << "   Saved to multi_demo.png\n\n";
}

void demo_histogram()
{
    std::cout << "📊 Histogram\n";
    std::vector<float> data;
    
    // Generate normal distribution data
    std::srand(123);
    for (int i = 0; i < 1000; ++i)
    {
        // Box-Muller transform for normal distribution
        float u1 = std::rand() / float(RAND_MAX);
        float u2 = std::rand() / float(RAND_MAX);
        float z0 = std::sqrt(-2.0f * std::log(u1)) * std::cos(2.0f * M_PI * u2);
        data.push_back(z0);
    }
    
    spectra::RenderOptions opts;
    opts.save_path = "histogram_demo.png";
    auto img = spectra::render_histogram(data, 30, opts);
    std::cout << "   Histogram of 1000 normal samples (30 bins)\n";
    std::cout << "   Saved to histogram_demo.png\n\n";
}

void demo_with_labels()
{
    std::cout << "📝 Plot with labels\n";
    std::vector<float> x = {0, 1, 2, 3, 4, 5};
    std::vector<float> y = {0, 1, 4, 9, 16, 25};
    
    spectra::RenderOptions opts;
    opts.title = "Quadratic Growth";
    opts.xlabel = "Time (seconds)";
    opts.ylabel = "Value";
    opts.save_path = "labeled_demo.png";
    
    auto img = spectra::render(x, y, opts);
    std::cout << "   Plot with title and axis labels\n";
    std::cout << "   Saved to labeled_demo.png\n\n";
}

void demo_performance()
{
    std::cout << "⚡ Performance test\n";
    
    // Large dataset
    const size_t n = 50000;
    std::vector<float> x, y;
    x.reserve(n);
    y.reserve(n);
    
    for (size_t i = 0; i < n; ++i)
    {
        x.push_back(i * 0.01f);
        y.push_back(std::sin(x[i] * 0.5f) + (std::rand() / float(RAND_MAX) - 0.5f) * 0.2f);
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    spectra::RenderOptions opts;
    opts.width = 1600;
    opts.height = 900;
    auto img = spectra::render(x, y, opts);
    auto end = std::chrono::high_resolution_clock::now();
    
    auto elapsed = std::chrono::duration<float>(end - start).count();
    std::cout << "   Rendered " << n << " points in " << elapsed << " seconds\n";
    std::cout << "   Image size: " << img.width << "x" << img.height 
              << " (" << img.size_bytes() << " bytes)\n";
    std::cout << "   Throughput: " << size_t(n / elapsed) << " points/second\n\n";
}

void demo_rendered_image()
{
    std::cout << "🖼️  RenderedImage object API\n";
    std::vector<float> x = {0, 1, 2, 3, 4};
    std::vector<float> y = {0, 1, 4, 9, 16};
    
    auto img = spectra::render(x, y);
    
    std::cout << "   RenderedImage attributes:\n";
    std::cout << "     .width = " << img.width << "\n";
    std::cout << "     .height = " << img.height << "\n";
    std::cout << "     .stride() = " << img.stride() << "\n";
    std::cout << "     .size_bytes() = " << img.size_bytes() << "\n";
    std::cout << "     .empty() = " << (img.empty() ? "true" : "false") << "\n";
    std::cout << "     .pixels() returns " << (void*)img.pixels() << "\n\n";
}

int main()
{
    std::cout << "🚀 Spectra Easy Embed Demo (C++)\n";
    std::cout << "=====================================\n\n";
    
    try
    {
        // Run all demos
        demo_basic_line();
        demo_save_to_file();
        demo_custom_size();
        demo_scatter();
        demo_multi_series();
        demo_histogram();
        demo_with_labels();
        demo_performance();
        demo_rendered_image();
        
        std::cout << "✅ All demos completed successfully!\n\n";
        
        // List generated files
        std::cout << "Generated files:\n";
        const char* files[] = {
            "easy_embed_demo.png",
            "scatter_demo.png", 
            "multi_demo.png",
            "histogram_demo.png",
            "labeled_demo.png"
        };
        
        for (const char* file : files)
        {
            std::ifstream f(file);
            if (f.good())
            {
                f.seekg(0, std::ios::end);
                auto size = f.tellg();
                std::cout << "   " << file << " (" << size << " bytes)\n";
            }
        }
        
        return 0;
    }
    catch (const std::exception& e)
    {
        std::cerr << "❌ Error: " << e.what() << "\n";
        std::cerr << "\nMake sure you built with Vulkan support:\n";
        std::cerr << "  cmake -DSPECTRA_BUILD_EMBED_SHARED=ON ..\n";
        return 1;
    }
}

File: python/examples/easy_embed_demo.py
#!/usr/bin/env python3
"""
Easy Embed Demo — One-liner offscreen rendering with Spectra.

Shows the new spectra.embed easy API for rendering plots to pixels or files
without any windows, daemon, or event loop. Perfect for:
- Web services (plot as PNG endpoint)
- Data pipelines (batch report generation)
- Scientific notebooks (embed plots in documents)
- Game engines (render UI overlays)

Build requirements:
    cmake -DSPECTRA_BUILD_EMBED_SHARED=ON ..
    export SPECTRA_EMBED_LIB=/path/to/build/libspectra_embed.so
"""

import math
import os
import sys
import tempfile
from pathlib import Path

# Ensure we can import spectra from the source tree
sys.path.insert(0, str(Path(__file__).parent.parent))

import spectra.embed as spe


def demo_basic_line():
    """Simple line plot to pixels."""
    print("📈 Basic line plot")
    x = [0, 1, 2, 3, 4, 5]
    y = [0, 1, 4, 9, 16, 25]  # y = x^2
    
    img = spe.render(x, y)
    print(f"   Rendered {img.width}x{img.height} image ({len(img)} bytes)")
    print(f"   Non-zero pixels: {sum(1 for b in img.data if b != 0)}")
    print()


def demo_save_to_file():
    """Render and save to PNG file."""
    print("💾 Save to PNG file")
    x = [0, 1, 2, 3, 4, 5]
    y = [0, 1, 4, 9, 16, 25]
    
    with tempfile.NamedTemporaryFile(suffix=".png", delete=False) as f:
        path = f.name
    
    spe.render(x, y, save=path)
    size = os.path.getsize(path)
    print(f"   Saved to {path} ({size} bytes)")
    
    # Clean up
    os.unlink(path)
    print()


def demo_custom_size():
    """Custom resolution."""
    print("🖼️  Custom size")
    x = [0, 1, 2, 3, 4, 5]
    y = [0, 1, 4, 9, 16, 25]
    
    img = spe.render(x, y, width=1920, height=1080)
    print(f"   HD render: {img.width}x{img.height}")
    print()


def demo_scatter():
    """Scatter plot."""
    print("🔵 Scatter plot")
    import random
    
    # Generate random scatter data
    random.seed(42)
    x = [random.random() * 10 for _ in range(100)]
    y = [random.random() * 10 for _ in range(100)]
    
    spe.scatter(x, y, save="scatter_demo.png")
    print("   Scatter plot with 100 points saved to scatter_demo.png")
    print()


def demo_multi_series():
    """Multiple series on one plot."""
    print("📊 Multi-series plot")
    import math
    
    x = [i * 0.1 for i in range(100)]
    y1 = [math.sin(v) for v in x]
    y2 = [math.cos(v) for v in x]
    y3 = [v * 0.5 for v in x]  # Linear
    
    spe.render_multi([
        (x, y1, "b-", "sin(x)"),
        (x, y2, "r--", "cos(x)"),
        (x, y3, "g:", "0.5x"),
    ], title="Trigonometric Functions", save="multi_demo.png")
    
    print("   Multi-series plot with sin, cos, and linear functions")
    print("   Saved to multi_demo.png")
    print()


def demo_histogram():
    """Histogram plot."""
    print("📊 Histogram")
    import random
    
    # Generate normal distribution data
    random.seed(123)
    data = [random.gauss(0, 1) for _ in range(1000)]
    
    spe.histogram(data, bins=30, save="histogram_demo.png")
    print("   Histogram of 1000 normal samples (30 bins)")
    print("   Saved to histogram_demo.png")
    print()


def demo_with_labels():
    """Plot with title and axis labels."""
    print("📝 Plot with labels")
    x = [0, 1, 2, 3, 4, 5]
    y = [0, 1, 4, 9, 16, 25]
    
    spe.render(x, y,
               title="Quadratic Growth",
               xlabel="Time (seconds)",
               ylabel="Value",
               save="labeled_demo.png")
    
    print("   Plot with title and axis labels")
    print("   Saved to labeled_demo.png")
    print()


def demo_numpy_support():
    """Show numpy array support."""
    print("🔢 NumPy array support")
    try:
        import numpy as np
    except ImportError:
        print("   NumPy not available, skipping demo")
        return
    
    # Create numpy arrays
    x = np.linspace(0, 4 * np.pi, 200)
    y = np.sin(x) * np.exp(-x / (4 * np.pi))  # Damped sine wave
    
    spe.render(x, y, save="numpy_demo.png")
    print("   Rendered damped sine wave from numpy arrays")
    print(f"   Array shapes: x={x.shape}, y={y.shape}")
    print("   Saved to numpy_demo.png")
    print()


def demo_performance():
    """Performance test with large dataset."""
    print("⚡ Performance test")
    import time
    import numpy as np
    
    # Large dataset
    n = 50000
    x = np.linspace(0, 100, n)
    y = np.sin(x * 0.5) + np.random.normal(0, 0.1, n)
    
    start = time.time()
    img = spe.render(x, y, width=1600, height=900)
    elapsed = time.time() - start
    
    print(f"   Rendered {n:,} points in {elapsed:.3f} seconds")
    print(f"   Image size: {img.width}x{img.height} ({len(img):,} bytes)")
    print(f"   Throughput: {n/elapsed:,.0f} points/second")
    print()


def demo_batch_processing():
    """Batch processing multiple plots."""
    print("📦 Batch processing")
    
    plots = [
        ("linear", lambda i: (list(range(10)), list(range(10)))),
        ("quadratic", lambda i: (list(range(10)), [j*j for j in range(10)])),
        ("cubic", lambda i: (list(range(10)), [j*j*j for j in range(10)])),
        ("sine", lambda i: (list(range(10)), [math.sin(j) for j in range(10)])),
    ]
    
    with tempfile.TemporaryDirectory() as tmpdir:
        for name, func in plots:
            x, y = func(0)
            path = os.path.join(tmpdir, f"{name}.png")
            spe.render(x, y, save=path)
            print(f"   Generated {name}.png")
        
        print(f"   All plots saved to {tmpdir}")
    print()


def demo_image_object():
    """Show Image object methods."""
    print("🖼️  Image object API")
    x = [0, 1, 2, 3, 4]
    y = [0, 1, 4, 9, 16]
    
    img = spe.render(x, y)
    
    print(f"   Image attributes:")
    print(f"     .width = {img.width}")
    print(f"     .height = {img.height}")
    print(f"     .stride = {img.stride}")
    print(f"     .size_bytes = {img.size_bytes}")
    print(f"     len(img) = {len(img)}")
    print(f"     bool(img) = {bool(img)}")
    print(f"     img.data is {len(img.data)} bytes of type {type(img.data)}")
    print()


def main():
    """Run all demos."""
    print("🚀 Spectra Easy Embed Demo")
    print("=" * 50)
    print()
    
    # Check if embed library is available
    try:
        # Try a minimal render to verify the library works
        spe.render([0, 1], [0, 1])
    except Exception as e:
        print(f" Error: {e}")
        print()
        print("Make sure you built the embed shared library:")
        print("  cmake -DSPECTRA_BUILD_EMBED_SHARED=ON ..")
        print("  export SPECTRA_EMBED_LIB=/path/to/build/libspectra_embed.so")
        print()
        return 1
    
    # Run demos
    demo_basic_line()
    demo_save_to_file()
    demo_custom_size()
    demo_scatter()
    demo_multi_series()
    demo_histogram()
    demo_with_labels()
    demo_numpy_support()
    demo_performance()
    demo_batch_processing()
    demo_image_object()
    
    print("✅ All demos completed successfully!")
    print()
    print("Generated files:")
    for f in ["scatter_demo.png", "multi_demo.png", "histogram_demo.png",
              "labeled_demo.png", "numpy_demo.png"]:
        if os.path.exists(f):
            size = os.path.getsize(f)
            print(f"   {f} ({size} bytes)")
    
    return 0


if __name__ == "__main__":
    sys.exit(main())
